libsolidity/codegen/ir/IRVariable.cpp

```diff
@@ -0,0 +1,90 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <libsolidity/codegen/ir/IRVariable.h>
+#include <libsolidity/ast/AST.h>
+#include <boost/range/adaptor/transformed.hpp>
+#include <libsolutil/StringUtils.h>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::frontend;
+using namespace solidity::util;
+
+IRVariable::IRVariable(VariableDeclaration const& _declaration):
+	m_type(*_declaration.annotation().type),
+	m_baseName("vloc_" + _declaration.name() + '_' + std::to_string(_declaration.id()))
+{
+	solAssert(!_declaration.isStateVariable(), "");
+}
+
+IRVariable::IRVariable(Type const& _type, std::string _name):
+	m_type(_type), m_baseName(std::move(_name))
+{
+}
+
+IRVariable::IRVariable(Expression const& _expression):
+	m_type(*_expression.annotation().type),
+	m_baseName("expr_" + to_string(_expression.id()))
+{
+}
+
+IRVariable IRVariable::part(std::string const& _slot, Type const& _type) const
+{
+	return IRVariable(_type, part(_slot));
+}
+string IRVariable::part(std::string const& _slot) const
+{
+	if (_slot.empty())
+		return m_baseName;
+	else
+		return m_baseName + '_' + _slot;
```

Comment:
Shouldn't this validate the slot name?

Reply:
It could - I was a bit afraid of the overhead of looking through the whole list of slots each time this is used, but it's probably fine - I'll add the check!